File: 029.Divide_Two_Integers/Solution.java
class Solution {
    public int divide(int dividend, int divisor) {
        int ans = -1;
        int sign = 1;
        if (dividend > 0) {
            sign = opposite(sign);
            dividend = opposite(dividend);
        }
        if (divisor > 0) {
            sign = opposite(sign);
            divisor = opposite(divisor);
        }   

        int origin_dividend = dividend;
        int origin_divisor = divisor;
        if (dividend > divisor) {
            return 0;
        } 

        dividend -= divisor;
        while (divisor >= dividend) {
            ans = ans + ans;
            divisor += divisor;
            dividend -= divisor;
        }
        int a = ans + opposite(divide(origin_dividend - divisor, origin_divisor));
        if(a == Integer.MIN_VALUE){
            if( sign > 0){
                return Integer.MAX_VALUE;
            }else{
                return Integer.MIN_VALUE;
            }
        }else{
            if(sign > 0){
                return opposite(a);
            }else{
                return a;
            }
        }
    }
    public int opposite(int x) {
        return ~x + 1;
    }
}

File: 029.Divide_Two_Integers/Solution.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        is_negative = False
        if (dividend > 0 and divisor < 0) or (dividend < 0 and divisor > 0):
            is_negative = True
        dividend = abs(dividend)
        divisor = abs(divisor)
        runner, index = divisor, 1
        track = []
        while runner <= dividend:
            track.append((runner, index))
            runner = runner + runner
            index += index
        ret = 0
        for index in range(len(track) - 1, -1, -1):
            val, idx = track[index]
            if dividend >= val:
                dividend -= val
                ret += idx
        if (not is_negative and ret >= 2147483648) or (is_negative and ret > 2147483648):
            return 2147483647
        return ret if not is_negative else -ret
